controllers/circuitrelay_controller.go

```diff
@@ -0,0 +1,388 @@
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/json"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	crypto "github.com/libp2p/go-libp2p-crypto"
+	clusterv1alpha1 "github.com/redhat-et/ipfs-operator/api/v1alpha1"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// CircuitRelayReconciler reconciles a CircuitRelay object
+type CircuitRelayReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=*,resources=*,verbs=get;list
+//+kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cluster.ipfs.io,resources=circuitrelays,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cluster.ipfs.io,resources=circuitrelays/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=cluster.ipfs.io,resources=circuitrelays/finalizers,verbs=update
+
+func (r *CircuitRelayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	instance := &clusterv1alpha1.CircuitRelay{}
+	if err := r.Get(ctx, req.NamespacedName, instance); err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			log.Info("CircuitRelay  resource not found. Ignoring since object must be deleted")
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		log.Error(err, "Failed to get CircuitRelay")
+		return ctrl.Result{}, err
+	}
+
+	svc := corev1.Service{}
+	svcMut := r.serviceRelay(instance, &svc)
+	_, err := controllerutil.CreateOrPatch(ctx, r.Client, &svc, svcMut)
+	if err != nil {
+		log.Error(err, "error during CreateOrPatch service", "name", instance.Name)
+		return ctrl.Result{
+			Requeue: true,
+		}, err
+	}
+
+	// Check that we have a public address
+	// if not, wait until it's available.
+	rsvc := corev1.Service{}
+	err = r.Get(ctx, client.ObjectKeyFromObject(&svc), &rsvc)
+	if err != nil {
+		log.Error(err, "error looking up service", "name", instance.Name)
+		return ctrl.Result{
+			Requeue: true,
+		}, err
+	}
+
+	if len(rsvc.Status.LoadBalancer.Ingress) == 0 {
+		log.Info("still waiting for service ingress addresses", "name", instance.Name)
+		return ctrl.Result{
+			RequeueAfter: time.Minute,
+		}, err
+	}
+
+	maddrs := make([]ma.Multiaddr, len(rsvc.Status.LoadBalancer.Ingress)*len(rsvc.Spec.Ports))
+	for i, addr := range rsvc.Status.LoadBalancer.Ingress {
```

Comment:
Converting from Kubernetes objects into multiaddrs will likely be important in the future, beyond the scope of IPFS Cluster.
We should move this functionality into a separate package such as `pkg/utils`.